File: src/main/java/net/arvandor/numinoustreasury/command/node/session/NodeSessionStaminaCommand.java
package net.arvandor.numinoustreasury.command.node.session;

import com.rpkit.core.service.Services;
import net.arvandor.numinoustreasury.NuminousTreasury;
import net.arvandor.numinoustreasury.node.NuminousNodeCreationSession;
import net.arvandor.numinoustreasury.node.NuminousNodeService;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.conversations.*;
import org.bukkit.entity.Player;

import java.util.List;

import static net.md_5.bungee.api.ChatColor.GRAY;
import static net.md_5.bungee.api.ChatColor.RED;

public final class NodeSessionStaminaCommand implements CommandExecutor, TabCompleter {

    private final ConversationFactory conversationFactory;

    public NodeSessionStaminaCommand(NuminousTreasury plugin) {
        conversationFactory = new ConversationFactory(plugin)
                .withModality(false)
                .withFirstPrompt(new ExperiencePrompt())
                .withEscapeSequence("cancel")
                .thatExcludesNonPlayersWithMessage(RED + "You must be a player to create nodes.")
                .addConversationAbandonedListener(event -> {
                    if (!event.gracefulExit()) {
                        Conversable forWhom = event.getContext().getForWhom();
                        if (forWhom instanceof Player player) {
                            player.sendMessage(RED + "Operation cancelled.");
                        }
                    }
                });
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (!sender.hasPermission("numinoustreasury.command.node.create")) {
            sender.sendMessage(RED + "You do not have permission to create nodes.");
            return true;
        }
        if (!(sender instanceof Player player)) {
            sender.sendMessage(RED + "You must be a player to create nodes.");
            return true;
        }
        NuminousNodeService nodeService = Services.INSTANCE.get(NuminousNodeService.class);
        NuminousNodeCreationSession session = nodeService.getNodeCreationSession(player);
        if (session == null) {
            session = new NuminousNodeCreationSession();
            nodeService.setNodeCreationSession(player, session);
        }
        if (args.length > 0) {
            int experience;
            try {
                experience = Integer.parseInt(args[0]);
            } catch (NumberFormatException exception) {
                sender.sendMessage(RED + "Stamina must be an integer.");
                return true;
            }
            session.setStaminaCost(experience);
            session.display(player);
        } else {
            Conversation conversation = conversationFactory.buildConversation(player);
            conversation.getContext().setSessionData("session", session);
            conversation.begin();
        }
        return true;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
        return List.of();
    }

    private static class ExperiencePrompt extends NumericPrompt {

        @Override
        public String getPromptText(ConversationContext context) {
            return GRAY + "Please enter the stamina cost when using the node:";
        }

        @Override
        protected Prompt acceptValidatedInput(ConversationContext context, Number input) {
            NuminousNodeCreationSession session = (NuminousNodeCreationSession) context.getSessionData("session");
            session.setStaminaCost(input.intValue());
            Conversable forWhom = context.getForWhom();
            if (forWhom instanceof Player player) {
                session.display(player);
            }
            return END_OF_CONVERSATION;
        }
    }

}

File: src/main/kotlin/net/arvandor/numinoustreasury/command/node/session/NodeSessionStaminaCommand.kt
package net.arvandor.numinoustreasury.command.node.session

import com.rpkit.core.service.Services
import net.arvandor.numinoustreasury.NuminousTreasury
import net.arvandor.numinoustreasury.node.NuminousNodeCreationSession
import net.arvandor.numinoustreasury.node.NuminousNodeService
import net.md_5.bungee.api.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.conversations.ConversationAbandonedEvent
import org.bukkit.conversations.ConversationContext
import org.bukkit.conversations.ConversationFactory
import org.bukkit.conversations.NumericPrompt
import org.bukkit.conversations.Prompt
import org.bukkit.entity.Player

class NodeSessionStaminaCommand(plugin: NuminousTreasury) : CommandExecutor, TabCompleter {
    private val conversationFactory: ConversationFactory

    init {
        conversationFactory =
            ConversationFactory(plugin)
                .withModality(false)
                .withFirstPrompt(ExperiencePrompt())
                .withEscapeSequence("cancel")
                .thatExcludesNonPlayersWithMessage(ChatColor.RED.toString() + "You must be a player to create nodes.")
                .addConversationAbandonedListener { event: ConversationAbandonedEvent ->
                    if (!event.gracefulExit()) {
                        val forWhom = event.context.forWhom
                        if (forWhom is Player) {
                            forWhom.sendMessage(ChatColor.RED.toString() + "Operation cancelled.")
                        }
                    }
                }
    }

    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): Boolean {
        if (!sender.hasPermission("numinoustreasury.command.node.create")) {
            sender.sendMessage(ChatColor.RED.toString() + "You do not have permission to create nodes.")
            return true
        }
        if (sender !is Player) {
            sender.sendMessage(ChatColor.RED.toString() + "You must be a player to create nodes.")
            return true
        }
        val nodeService =
            Services.INSTANCE.get(
                NuminousNodeService::class.java,
            )
        var session = nodeService.getNodeCreationSession(sender)
        if (session == null) {
            session = NuminousNodeCreationSession()
            nodeService.setNodeCreationSession(sender, session)
        }
        if (args.size > 0) {
            val experience: Int
            try {
                experience = args[0].toInt()
            } catch (exception: NumberFormatException) {
                sender.sendMessage(ChatColor.RED.toString() + "Stamina must be an integer.")
                return true
            }
            session.staminaCost = experience
            session.display(sender)
        } else {
            val conversation = conversationFactory.buildConversation(sender)
            conversation.context.setSessionData("session", session)
            conversation.begin()
        }
        return true
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): List<String>? {
        return listOf()
    }

    private class ExperiencePrompt : NumericPrompt() {
        override fun getPromptText(context: ConversationContext): String {
            return ChatColor.GRAY.toString() + "Please enter the stamina cost when using the node:"
        }

        override fun acceptValidatedInput(
            context: ConversationContext,
            input: Number,
        ): Prompt? {
            val session = context.getSessionData("session") as NuminousNodeCreationSession
            session.staminaCost = input.toInt()
            val forWhom = context.forWhom
            if (forWhom is Player) {
                session.display(forWhom)
            }
            return END_OF_CONVERSATION
        }
    }
}
